File: src/CachedFib.java
import java.util.HashMap;
import java.util.function.Function;

public class CachedFib {
    //Attempt at making a closure work for caching.


    public static Function<Integer, Integer> fib() {

        HashMap<Integer, Integer> map = new HashMap<>();

        return new Function<Integer, Integer>() {
            @Override
            public Integer apply(Integer n) {
                {
                    if (map.containsKey(n)) {
                        return map.get(n);
                    }

                    if (n < 2) {
                        System.out.println("Putting " + n + " into map as: " + n);
                        map.put(n, n);
                        return n;
                    }
                    ;

                    int result = this.apply(n - 1) + this.apply(n - 2);
                    System.out.println("Putting " + n + " into map as: " + result);
                    map.put(n, result);
                    return result;
                }
            }
        };
    }

    public static void main(String[] args){
        Function<Integer, Integer> cachedFib = fib();
        System.out.println(cachedFib.apply(5));
        /**
         * Output:
         Putting 1 into map as: 1
         Putting 0 into map as: 0
         Putting 2 into map as: 1
         Putting 3 into map as: 2
         Putting 4 into map as: 3
         Putting 5 into map as: 5
         5
         */
    }
}

File: src/CachedFib.py
def cache(func):
    cache = {}

    def wrapper(*args):
        if args not in cache:
            result = func(*args)
            print("putting " + str(*args) + " in cache as " + str(result))
            cache[args] = result
        return cache[args]

    return wrapper


@cache
def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)


def main():
    print(fib(5))


if __name__ == '__main__':
    main()

# Output:
# putting 1 in cache as 1
# putting 0 in cache as 0
# putting 2 in cache as 1
# putting 3 in cache as 2
# putting 4 in cache as 3
# putting 5 in cache as 5
# 5
